implement instant doubles functionality (clone deck)

iamcodemaker opened the pull request:
This is a start on implementing instant doubles functionality. As is, copying rate, and loop state works. Copying play state works sometimes. Copying pitch_adjust and playposition don't seem to work at all. I'm open to suggestions.

https://bugs.launchpad.net/mixxx/+bug/408111

iamcodemaker committed:
add method to copy playback state from another deck

This will be used to implement instant doubles.

iamcodemaker commented on src/test/autodjprocessor_test.cpp:
```diff
@@ -77,6 +77,10 @@ class FakeDeck : public BaseTrackPlayer {
 
+    void slotCopyFrom(const QString& group) override {
+        return;
+    }
```
I'm not sure what should go here.

Be-ing replied:
[MOCK_METHOD1](https://github.com/google/googletest/blob/master/googlemock/docs/ForDummies.md#writing-the-mock-class)?

iamcodemaker committed:
perform copy operation after track is loaded

iamcodemaker commented on src/mixer/basetrackplayer.cpp:
```diff
@@ -343,22 +345,58 @@ void BaseTrackPlayerImpl::slotTrackLoaded(TrackPointer pNewTrack,
+            // copy the loop state
+            pControl = ControlObject::getControl(m_copyFrom, "loop_enabled");
+            if (pControl) {
+                m_pLoopEnabled->set(pControl->get());
```
this makes the loop get highlighted wavform and the button lights up, but it doesn't actually loop

iamcodemaker commented on src/mixer/basetrackplayer.cpp:
```diff
@@ -343,22 +345,58 @@ void BaseTrackPlayerImpl::slotTrackLoaded(TrackPointer pNewTrack,
+            pControl = ControlObject::getControl(m_copyFrom, "playposition");
+            if (pControl) {
+                m_pChannel->getEngineBuffer()->slotControlSeekExact(pControl->get());
+            }
```
this seems to have no effect

daschuer replied:
There is somewhere code that seeks to the cue point or 0.0 on track load.
Maybe this happens afterwards.

daschuer replied:
I also think we have to move this code into the engine to allow an sample exact seek.
It may happen that the Playposition co is outdated. So you have to calculate the future value. You need also consider the processing order of the decks inside the engine.
There is already something similar in visualPlayposition that calculates the exact position for rendering the waveforms.

iamcodemaker replied:
I modified this to set `"playposition"` since the value from `"playposition"` is fractional, but the value that `slotControlSeekExact()` accepts is not. This now sets the play position correctly, but the copied track ends up slightly behind as you suspected it would.

iamcodemaker commented on src/mixer/basetrackplayer.cpp:
```diff
@@ -343,22 +345,58 @@ void BaseTrackPlayerImpl::slotTrackLoaded(TrackPointer pNewTrack,
+            pControl = ControlObject::getControl(m_copyFrom, "pitch_adjust");
+            if (pControl && m_pPitchAdjust != NULL) {
+                m_pPitchAdjust->set(pControl->get());
+            }
```
this works, but only if the other deck was playing when the track was copied

daschuer replied:
You may add a qDebug() here to check the current value of the others track pitch adjust. I have no idea why this can be 0.0 but we will find out.

iamcodemaker reviewed (commented):
Some of the elements aren't copied. I'm not sure where to go from here. Any ideas?

daschuer commented on src/mixer/basetrackplayer.cpp:
```diff
@@ -343,22 +345,58 @@ void BaseTrackPlayerImpl::slotTrackLoaded(TrackPointer pNewTrack,
+            // copy play state
+            pControl = ControlObject::getControl(m_copyFrom, "play");
+            if (pControl) {
+                m_pPlay->set(pControl->get());
```
Please set the play state at the very end.
Else the track may reach loop end before setting the loop.
I can remember also an issue with loading and playing a track at the same time.
I need to have a closer look tonight.

daschuer reviewed (commented):
Thank you for the work so far. I assume your problems are mainly cases by follow up code reverting your changes. So I think the code need to be moved arround a bit and special case these places losing the initial or the copied values.
I think in general, you shools load a paused track, adjust the COs and then play and seek it with special engine code that achieved playing sample exact like the source track. If this is a case we will get +6 dB volume. If not we get an unwanted phaser effect.

iamcodemaker committed:
rework instant doubles copying

iamcodemaker commented on src/mixer/basetrackplayer.cpp:
```diff
@@ -343,22 +343,44 @@ void BaseTrackPlayerImpl::slotTrackLoaded(TrackPointer pNewTrack,
+            }
+
+            // copy the play position
+            ControlObject::set(ConfigKey(getGroup(), "playposition"), ControlObject::get(ConfigKey(m_copyFrom, "playposition")));
```
this works, but it does a standard seek which is affected by quantize. I didn't test it, but I'm guessing that will mess everything up. Also the track is slightly behind the playing track. I'll look into moving this set into the engine somehow, I still need to wrap my head around what is running in which threads and how to get the necessary data to the right place.

daschuer replied:
That can't be the right solution. Since the play position is the position of the source track after the last engine call.
If there are engine calls in between, adopting the new Playposition at the target track it will be behind.
To avoid this you need to patch the engine code.
I think you may use an atomic varianle, or an atomic struct using ControlValueAtomic
That stores tells the engine:
"Copy the source Playposition"
"Adjust it by one buffer size in case the destination deck is has a higher index"
"Check these buffer in the destination track EngineBuffer" 
"Adopt it without quantize"

The trigger can be set here.
The actual copy needs to be done in the engine thread, in the EngineBuffer::process() call.

iamcodemaker replied:
I'll look into it. I think I know how to make it work, I just need to figure out how to get the play position from one enginebuffer to the other.

iamcodemaker commented:
With the latest revision, this works, but the play position is slightly behind.

daschuer commented:
slotControlSeekExact() excepts double, there must be another problem.

daschuer commented:
If you move thelogic to EngineBuffer:: process, you can access all data. You just need a static atomic varianle that is used to communicate between the instances if EmgineBuffer. Did you have a look at control valve atomic? IMHO that suites well as a container class for all shared data.

rryan commented:
> If you move thelogic to EngineBuffer:: process,

To be specific, please move just the "sync to other deck's playposition atomically" logic, not the whole instant doubles functionality! Let's try to keep as much "business logic" as possible out of the engine, as it's already a rats nest.

rryan commented:
> `"playposition"` is a fractional value between
> 0.0 and 1.0 but `slotControlSeekExact()` does not expect a fractional value.

You can get the "track_samples" CO to get the number of samples in a deck's loaded track -- you can multiply that by playposition to get a sample value for slotControlSeekExact

iamcodemaker commented:
Ahh, I wondering if you could get the non fractional position. Thanks.

daschuer commented:
I have looked into the engine code when working on https://github.com/mixxxdj/mixxx/pull/1896
I think your task is not too hard.
I think you may tweak BpmControl::getNearestPositionInPhase()
To use the copy source as sync target.
Than you need to handle the case where no beatgrid is available, and add a new seek the calling this function.

iamcodemaker commented:
I'll look at that. I have been looking at adding code to EngineMaster that tells one channel to seek to the position of another. I need to figure out the best way to trigger it though.

iamcodemaker committed:
perform the seek operation in the engine thread

iamcodemaker commented:
I pushed a new rev that performs the copy operation in the engine thread. This works perfectly, the new track is started in the exact right position. I am going to work on cleaning this up. I think I am going to use `EngineSync::pickNonSyncSyncTarget()` to find the source channel and then tell `EngineMaster` to clone the position. Or cloning the sync target's position may need to be a special seek request or a special sync request. I think it kinda makes sense as a sync request.

iamcodemaker committed:
use EngineSync::pickNonSyncSyncTarget() to determine the copy source

Also queue the position clone using a special sync request.

iamcodemaker commented:
I cleaned this up some. Now it uses `EngineSync::pickNonSyncSyncTarget()` to determine which deck to copy from. Also the play position copy is triggered from a special sync request I added to `EngineBuffer`. I plan to add code to do a copy instead of a normal load if the same track is loaded twice withing a certain period of time. I am also going to look into adding some tests.

There are still some oddities around looping. It looks like the loop in and out points aren't copied to duplicate track. I'll need to set them manually.

iamcodemaker committed:
add comment

iamcodemaker committed:
when cloning, copy loop info from the target track

iamcodemaker commented:
updated the loop copying code

iamcodemaker commented on src/engine/enginebuffer.cpp:
```diff
@@ -430,6 +430,11 @@ void EngineBuffer::requestSyncMode(SyncMode mode) {
 
+void EngineBuffer::requestSyncPosition() {
+    m_iEnableSyncQueued = SYNC_REQUEST_POSITION;
+    m_iSyncModeQueued = SYNC_INVALID;
```
using two atomic variables in `EngineBuffer` seems like a race condition. This should be replaced by two vars in a struct with an atomic guard. Thoughts?

iamcodemaker committed:
fix autodjprocessor_test.cpp

daschuer reviewed (changes requested):
Nice progress. Thanks

I wonder if this should be the default if you drag a track from one deck to an other.

iamcodemaker committed:
whitespace

iamcodemaker committed:
perform a clone operation instead of a load on double tap

If two load operations are attempted in quick succession, a clone will
be performed instead of the second load.

iamcodemaker committed:
use getVisualPlayPos() when cloning play pos from other buffers

iamcodemaker committed:
don't reset m_iSyncModeQueued when copying position

iamcodemaker committed:
wrap long lines

iamcodemaker committed:
reset m_cloneFromChannel on track load failure

iamcodemaker committed:
explicitly pass the target channel when cloning a buffer

iamcodemaker committed:
specify the source channel when cloning a deck

iamcodemaker committed:
copy play state closer to the position sync operation

Previously the track would start playing as soon as it was loaded but
before the position was synced. Some audio would leak out on occasion.
Moving these operations closer together reduces this possibility. To
completely eliminate it, the seek and the play will need to happen in
the engine thread together. Not sure if this will be necessary.

iamcodemaker commented on src/engine/enginebuffer.cpp:
```diff
@@ -441,6 +446,10 @@ void EngineBuffer::readToCrossfadeBuffer(const int iBufferSize) {
 }
 
+void EngineBuffer::seekCloneBuffer(EngineBuffer* pOtherBuffer) {
+    doSeekPlayPos(pOtherBuffer->m_filepos_play, SEEK_EXACT);
```
I can do that.

iamcodemaker commented on src/engine/enginebuffer.cpp:
```diff
@@ -1137,6 +1147,12 @@ void EngineBuffer::processSyncRequests() {
         m_pEngineSync->requestEnableSync(m_pSyncControl, false);
         break;
+    case SYNC_REQUEST_POSITION:
+        pChannel = m_pEngineSync->pickNonSyncSyncTarget(m_pSyncControl->getChannel());
```
A `pickNonSyncSyncTarget()` call is still in there but, everything has been reworked to copy from a channel passed in. I'll need some guidance on how to determine which channel to copy from though.

iamcodemaker commented:
Ok, I think I have addressed the comments thus far. There are still some tweaks that need to be made though.

iamcodemaker commented on src/mixer/basetrackplayer.cpp:
```diff
@@ -245,6 +258,18 @@ void BaseTrackPlayerImpl::disconnectLoadedTrack() {
+    mixxx::Duration elapsed = m_cloneTimer.restart();
+    if (elapsed < mixxx::Duration::fromSeconds(1)) {
+        // load pressed twice quickly, clone instead of loading
+        EngineChannel* pChannel = m_pEngineMaster->getEngineSync()->pickNonSyncSyncTarget(m_pChannel);
```
this needs to be replaced with something more specific and predictable, but I don't know what that is. I'll need some guidance here.

iamcodemaker commented on src/mixer/basetrackplayer.cpp:
```diff
@@ -343,22 +369,50 @@ void BaseTrackPlayerImpl::slotTrackLoaded(TrackPointer pNewTrack,
+
+            // copy play state
+            ControlObject::set(ConfigKey(getGroup(), "play"),
+                ControlObject::get(ConfigKey(m_cloneFromChannel->getGroup(), "play")));
```
previously I was triggering a play on load, but this caused the track to play before it was seeked and this was audible sometimes. Triggering the play here reduces the gap between play and seek, but it might make sense to move the play operation as part of the `requestSyncPosition()` operation though.

iamcodemaker replied:
I attempted to play from the engine thread as part of the sync operation, but it didn't work quite right.

Be-ing commented:
> I wonder if this should be the default if you drag a track from one deck to an other.

Yes, I think that would be the most straightforward way to expose the functionality in the GUI.

iamcodemaker commented:
I'll look into that.

iamcodemaker committed:
add slotCloneDeck to BaseTrackPlayer

This allows cloning the given deck identified by a string.

iamcodemaker committed:
perform a deck clone when dragging a track from one deck to another

Instead of just loading the track that was dragged and dropped, clone
the play state, play position, pitch/rate, and key.

iamcodemaker committed:
add function to get the exact play position

iamcodemaker committed:
use new getExactPlayPos() function

iamcodemaker commented on src/engine/enginebuffer.h:
```diff
@@ -144,6 +145,7 @@ class EngineBuffer : public EngineObject {
     bool isTrackLoaded();
     TrackPointer getLoadedTrack() const;
 
+    double getExactPlayPos();
```
wasn't sure of the naming here

iamcodemaker replied:
there is a `getVisualPlayPos`, maybe exact isn't the right term, but I think just naming it `getPlayPos` isn't as clear as it could be

Be-ing replied:
:man_shrugging:

iamcodemaker reviewed (commented):
I added a common function to get the play pos (visual pos*track samples). I also experimented with playing the track from the engine thread, but this caused the cloned track to be slightly ahead of where it should be using the getExactPlayPos().

Be-ing commented:
Any change in behavior is [going to disrupt someone's workflow](https://xkcd.com/1172/). I don't think any extra steps beyond dragging and dropping should be required for this. You can avoid doubling the level by lowering the volume fader on the deck being dropped onto before dropping the track.

daschuer commented:
https://www.youtube.com/watch?v=qLwsFQK4J6c
Serator: A preference option that enables instant double for drag and drop. 
Traktor: You need to drag and drop twice for the instant double feature.

iamcodemaker commented:
So do we need a config option for this? I think just making drag and drop do instant doubles is a reasonable choice. I feel like a config option is unnecessary. Once you learn the behavior, it should be fine.

Be-ing commented:
I agree. I don't think there is a need for yet another preference option for this.

daschuer commented:
Yes. There is a risc for a +6 dB offset, but there should bee a steep learning curve to mute the target deck before. Currently I can't think of a use case where I don't want this the instant play feature when dragging from deck to deck.
This is different when dragging from or to samplers preview decks. The instant double feature can there crate odd situations because not all controls are acessable.

I think I also do not want this feature when dragging from the preview deck. 

What do you think?

iamcodemaker commented:
> This is different when dragging from or to samplers preview decks. The instant double feature can there crate odd situations because not all controls are acessable.

> I think I also do not want this feature when dragging from the preview deck.

> What do you think?

I agree, and currently it should only trigger from actual decks (though I didn't test that, I'll put it on my to-do list).

poelzi commented:
Even if I'm careful and have the volume slider down: If pitch lock is activated, starting another pitch locked deck is a very dangerous situation, even on my notebook. You might easily loose buffers due to latency spikes. I need a way to bypass this dangerous but useful feature and would like to have it only activated on conscious thought (pressing key).

daschuer commented:
That's a good point. 

For me it sounds very attractive to have instant double enabled by default for a normal d'n'd action.

What would be a natural way around it?

We may use the shift key known as "Move File" to disable instant double. See also this:
https://ux.stackexchange.com/questions/83748/what-are-the-most-common-modifier-keys-for-dragging-objects-with-a-mouse

An other option would be to use the quantize button to enable or disable instant double feature.

Or open a menu if key lock is involved.

The latest option seems to be the best discoverable option. We may add a "don't ask again" checkbox. Than we are back at a preference option though. :-/

... use Quantize?

Ideas?

daschuer commented on src/engine/enginebuffer.h:
```diff
@@ -387,6 +392,7 @@ class EngineBuffer : public EngineObject {
     QAtomicInt m_iEnableSyncQueued;
     QAtomicInt m_iSyncModeQueued;
     ControlValueAtomic<double> m_queuedSeekPosition;
+    ControlValueAtomic<EngineChannel*> m_pSyncPosition;
```
You can use here the qt class 
QAtomicPointer
The variable name is missleading. How about. m_pChannelForSync or something?

daschuer commented on src/engine/enginebuffer.cpp:
```diff
@@ -1144,6 +1152,11 @@ void EngineBuffer::processSyncRequests() {
         m_pEngineSync->requestSyncMode(m_pSyncControl,
                                        static_cast<SyncMode>(mode_request));
     }
+    EngineChannel* pChannel = m_pSyncPosition.getValue();
```
I wonder if we can't check this at the very beginning of this funktion?

Be-ing commented:
> We may use the shift key known as "Move File" to disable instant double.

I like this idea best. No preference option is required. You are right that it is not easy to discover, but I'm doubtful many users will be searching for it. Nevertheless, if we decide to do this we should document it in the manual.

poelzi commented:
A common and good way would be to show the action through the cursor symbol. We need two symbols, one for playing and one for not playing. Changing the shift key while dragging also changes symbol.

This is who dolphin, explorer and many other tools do it.

iamcodemaker committed:
use QAtomicPointer and rename variable

iamcodemaker committed:
only clone when doing drag and drop if the shift key is pressed

iamcodemaker commented:
I implemented holding shift while dropping to do the clone. My reasoning is this is the least surprising option. Users have to know its there to use it, but it doesn't change existing expectations.

The implementation will also do a clone if the user rapidly presses the load button twice. I think this is a better way to implement cloning than drag and drop. It works from controllers and keyboards without any modifications to the mapping and this is how both Serato and Rekordbox trigger instant doubles. 

Currently the double load clone uses `pickNonSyncSyncTarget` to decide which channel to clone. I have been told this can choose an unexpected target, so I'd like some input on a better way to pick the deck to clone from (maybe track the last deck that was played and always use that?). I think `pickNonSyncSyncTarget` is also used for when the user presses the `sync` button though, and that seems to work fine, so maybe this isn't a problem.

iamcodemaker committed:
change double load timeout to half a second

ronso0 commented:
Thinking about this with TerminalMix2 & 4 in mind (they have a Load button for each deck), I guess another way to clone a deck to another would be to hold the Load button of to-clone deck, then do a regular short-press on the deck that should receive the clone.
Are there any new controls so this can be done individually in a script?

iamcodemaker commented:
The Mixtrack Platinum I tested this on also has a load button for each deck. The double press is how serato and rekordbox do it. I don't think the hold and press is a bad idea, but it won't be what users expect.

I didn't add any new controls to do it from a script. That should be trivial once the other details are ironed out.

ronso0 commented:
I'm just wondering: do we really always need pick up folks from Serato, Traktor, [any other proprietary DJ s/w] when we introduce a similiar feature (maybe in even in a better way) just because almost any controller out there is labeled ahhh not for Mixxx?
I'm confident Mixxx can "walk on his own feet now" by now and we could think up a better solution (especially regarding the Sync pitfall you mentioned above). Sure, for any 2-deck controller, longpress or double-press would work because..well there's only 2 decks. But imagining the 4-channel controller scenario, I'd really love to see a physical counterpart to drag'n'drop.

ronso0 commented:
this came to mind because that's how hotcue preview works:
- press and hold (previously set) hotcue_N
- press Play to continue playing after hotcue release

see my wishlist bug lp:1811300 for drag'n'drop equivalent:
- klick and hold hotcue button to play from there
- keep holding, move cursor above Play button, release
= deck keeps playing

iamcodemaker commented:
Nothing really stands on its own. You also should consider people watching YouTube tutorials where this feature is used. First instinct will be to try the same thing on their Mixxx setup, not reach for the documentation.

As for the problem of selecting the playing deck, if only one other deck is playing, it's trivial to select the correct deck. If multiple decks are playing, the user just needs to know what to expect. I'm not sure how Serato/Rekordbox handle it, but cloning the deck from the opposite side of the controller would be one way to handle it (my preferred solution). Another way would be always cloning the last deck that was played. Making it predictable makes it so the user can accurately signal their intention and get the result they want.

ronso0 commented:
> Nothing really stands on its own. You also should consider people watching YouTube tutorials where this feature is used. First instinct will be to try the same thing on their Mixxx setup, not reach for the documentation.

Good point, but only halve a point IMO :wink: the videos I found focus on how to set up this feature in the preferences of a _specific software_, actually I found only one video that explains how do to it with a controller (yes, double-tap Load in that case). Maybe it's because this feature is most valuable to DJs having only one physical deck and no MIDI controller, how don't mind resorting to mouse/trackpad?

It's clear we need more Mixxx tutorials, especially regarding special features. Apparently, most of the power users aren't that video-oriented..or just busy with coding and testing.

Also interesting aspect: some software (Traktor?) has the option to automatically trigger instant double when loading a currently *again* to another deck. Especially for the one-physical-deck context, that's very comfortable: right after the transition is done, load that track again, switch over, and look for a follow-up track.
A preferences option would be very helpful for that.

Btw: AFAICT "Instant doubles" is a Serato name. Should we come up with something different?
Clone deck?
Duplicate deck?

Be-ing commented:
> I implemented holding shift while dropping to do the clone. My reasoning is this is the least surprising option. Users have to know its there to use it, but it doesn't change existing expectations.

I do not like this. It is not discoverable and it requires an extra step for something that I think users will want to do in a hurry while mixing. IMO instant doubles should be the default and holding shift while dropping could skip copying the play state to accommodate @poelzi's concern.

Be-ing commented:
> The implementation will also do a clone if the user rapidly presses the load button twice. I think this is a better way to implement cloning than drag and drop. It works from controllers and keyboards without any modifications to the mapping and this is how both Serato and Rekordbox trigger instant doubles.

I don't think this is a great way to implement this and I agree with @ronso0 that we shouldn't follow other software if we can do it better. This can make sense for two deck controllers, but with 4 decks there is no good way to determine which deck to either copy or load to, which can lead to unwelcome surprises during a performance. 

> Thinking about this with TerminalMix2 & 4 in mind (they have a Load button for each deck), I guess another way to clone a deck to another would be to hold the Load button of to-clone deck, then do a regular short-press on the deck that should receive the clone.

This is similar to how I implemented loading for the Xone K2 4 deck mapping. The Xone K2 has no dedicated load button, so I pressing the browse encoder load the first stopped deck. However, if the user holds the browse encoder pressed, they can select the deck to load to by pressing its play button before releasing the encoder. I find this works well.

I am thinking it may be better to hold the load button on the deck being loaded to, not the deck being copied from. In the worst case scenario when the user changes their mind before completing the action, they can release the load button and the selected track in the library would be loaded without starting playback, so there would be no disruption of the performance. However, if the user has the playing deck protection feature disabled in the preferences and they have to hold the load button on the deck being copied from, there's no way out without disrupting the performance between pressing & releasing the button.

Be-ing commented:
I wouldn't have suggested implementing the double press LoadSelectedTrack for instant doubles feature in the first place, but it does have the nice advantage of being usable without requiring any modification of controller mappings and it works for two deck controllers. But I think we still need a new ControlObject to allow mappings to explicitly specify which deck to copy from and which to load to. Should this new CO be [ChannelX], LoadFromDeck or [Channel1], CloneToDeck? Intuitively I am leaning towards [ChannelX], LoadFromDeck, I think because it is similar to the existing LoadSelectedTrack.

Be-ing commented:
> AFAICT "Instant doubles" is a Serato name. Should we come up with something different?
Clone deck?
Duplicate deck?

I like "clone". I don't think "instant doubles" is very self-explanatory.

ronso0 commented:
> However, if the user has the playing deck protection feature disabled in the preferences and they have to hold the load button on the deck being copied from, there's no way out without disrupting the performance between pressing & releasing the button.

Hmmkay I didn't consider that setting. Isn't that protection enabled the default? If not, why? Anyone seriously performing for others should have this enabled as stopping your main deck is pretty much the worst case in a live performance, and it can easily happen if you're in a hurry, nervous or just distracted for a second, or all at once.

Furthermore, IF the entire clone procedure is hardcoded or in components.js, there should be a timeout for long presses of Load IMO so a _long_ press would do nothing at all; half a second should cover that period of changing one's mind.

If we agree on the long-press & press procedure, I have strong preference for
[long-press Load on source deck] + [short-press Load on target deck]
 (resembles pick'n'place aka drag'n'drop)
vs.
[long-press target deck] + [press source deck]
 ('open' deck, then pick source)
as it doesn't introduce another way of picturizing the load procedure.

Besides, the long-press procedure, anything else comes to mind?

ronso0 commented:
> > I implemented holding shift while dropping to do the clone. My reasoning is this is the least surprising option. Users have to know its there to use it, but it doesn't change existing expectations.
> 
> I do not like this. It is not discoverable and it requires an extra step for something that I think users will want to do in a hurry while mixing. IMO instant doubles should be the default and holding shift while dropping could skip copying the play state to accommodate @poelzi's concern.

Both perspectives are valid:
we need to make the feature discoverable (> clone deck = default d'n'd action), and we should not disturb established (workflows > modifier required for cloning)
Until now, copying tracks was only possible via mouse d'n'd. What about showing a warning when this d'n'd is for done the first time after introducing instant doubles? This would not interrupt the controller workflow as the user is already on the mouse. Also we have such warnings for  unconfigured vinyl control, passthrough and diverse load failures (i.e. non-existent or unreadable tracks).
_______________
Mixxx can create a playing clone when you drop a currently playing track to another deck.
Do you want this to be the default action, or should Mixxx just load the track without
playing it? You can change this any time [Preferences link]
**[Load & Stop]   [Clone & Play]**
_______________

Thunar (xfce file manager) shows a tiny pop-up when drag'n'dropping with right mouse button:
![clone_popup](https://user-images.githubusercontent.com/5934199/51608388-9f5ea680-1f17-11e9-878a-2dbf009267cb.png)
In conjunction with "Always do this" checkbox this could work well IMO

Be-ing commented:
> Isn't that protection enabled the default?

Yes, it is.

>  there should be a timeout for long presses of Load IMO so a long press would do nothing at all

Eh, I'm not so enthusiastic about cancelling normal functionality if the user takes a long time. We could do it like I have mapped the browse encoder on the Xone K2 where loading isn't done until the encoder is released and the alternate behavior (in this case, deck cloning) is only done if another is pressed before button up.

Be-ing commented:
> we should not disturb established (workflows > modifier required for cloning)

https://xkcd.com/1172/
Let's not overcomplicate the design just for the sake of requiring someone not to press a modifier key.

ronso0 commented:
> https://xkcd.com/1172/

Nice one..

> Let's not overcomplicate the design just for the sake of requiring someone not to press a modifier key.

overcomplicated, really? a one-time user interaction to balance dicoverability & workflow.
however, the learning curve will be quite steep when cloning a deck to another with Vol fader up.. ;) which is bad practice anyway.

daschuer commented:
What may a user expect if he loads a track twice? Nothing? So this action is fee for instant doubles.

The same is for pressing load at a playing deck (if disabled) this is also free.

I think we can effort to have loading a playing deck disabled, for the controller trick.

If this is not the case the user can load the same track from the library twice, to have this feature enabled. It would be Ok for me, because the track is probably not far out of sight and else he can use the mouse.

The double load feature hase the advantage to not require two hands.

Be-ing commented:
> overcomplicated, really?

Yes, pop up dialogs are annoying, especially when mixing. IMO they should only be used when it's really important to interrupt the user, for example confirming before permanently deleting a file.

> the learning curve will be quite steep when cloning a deck to another with Vol fader up

I'm not so sure about that. It might not be so obvious to people who are completely unfamiliar with audio gear, but I think it will be fairly straightforward to figure out for experienced DJs, which is who I think will be most interested in this feature.

ronso0 commented:
> What may a user expect if he loads a track twice? Nothing? So this action is fee for instant doubles.

The use case @poelzi desccribes, fixing the beatgrid if you need it for the next quantized transition. But yeah, this is probably a rare case and we don't necessarily need to take it into account.

> The double load feature hase the advantage to not require two hands.

..two fingers ;)

ronso0 commented:
> > overcomplicated, really?
> 
> Yes, pop up dialogs are annoying, especially when mixing.

I was suggesting a one-time popup to propose available options just once. No-one would go live right after an update or fresh install without testing at least a basic routine.

> > the learning curve will be quite steep when cloning a deck to another with Vol fader up
> 
> I'm not so sure about that. It might not be so obvious to people who are completely unfamiliar with audio gear, ...

Especially those users will remember the moment when they clone a deck and the Master volume suddenly goes up :smile: quick learning (again: preparing a deck with Vol fader up is not wise at all)

So, yeah, let's go with 'clone deck' as default action.

daschuer commented:
I like the idea to have this default (until we get  serious complains). However AutoDJ needs to bypass this to avoid messing it up unattended.

Be-ing commented:
How does this interfere with AutoDJ?

daschuer commented:
If you have for some reasons two times the same track in the queue.
The second load must not be interpreted as instant doubles. Else both tracks will end at the same time. That's all.

Be-ing commented:
What is required for merging this? I think:
  * [ ] Resolve merge conflicts with master
  * [ ] Make cloning the default drag-and-drop action. Shift + drop clones everything except the play button state.
  * [ ] Add [ChannelX], CloneToDeck or [ChannelX], CloneFromDeck ControlObject -- we still need to decide which to implement.


> If you have for some reasons two times the same track in the queue.
The second load must not be interpreted as instant doubles. Else both tracks will end at the same time. That's all.

I don't think that would be an issue because the double load for cloning only happens when the two loads are done within 0.5 seconds.

Be-ing commented:
@iamcodemaker will you have time to finish this in the next week or so? If not, maybe someone else can pick it up so we can get this in 2.3 beta.

iamcodemaker commented:
In the next week? Depends on how difficult it is to implement ControlObjects. I'll take a look tonight. The other stuff is trivial. For the clone, we could implement both CloneTo and CloneFrom I think.

daschuer commented:
> For the clone, we could implement both CloneTo and CloneFrom I think.

One would be enough.This produces already 68*68 4624 Controls if we consider samplers.

Since no one wants to learn such a control, we can also consider an API like CO.
Like [Master], clone = From deck * 256 + Do deck for cloning.

Be-ing commented:
I agree, it would be superfluous to implement both CloneFromDeck and CloneToDeck. I think I'd prefer CloneFromDeck, but ultimately I think it's arbitrary.

> Like [Master], clone = From deck * 256 + Do deck for cloning.

I think that would be overcomplicated.

Be-ing commented on src/mixer/basetrackplayer.cpp:
```diff
@@ -36,7 +36,8 @@ BaseTrackPlayerImpl::BaseTrackPlayerImpl(QObject* pParent,
           m_pLoadedTrack(),
-          m_replaygainPending(false) {
+          m_replaygainPending(false),
+          m_cloneFromChannel(NULL) {
```
`NULL` is a legacy hack from before C++11. Please replace all new uses of `NULL` with `nullptr`

Be-ing commented on src/mixer/basetrackplayer.cpp:
```diff
@@ -168,21 +171,31 @@ void BaseTrackPlayerImpl::loadTrack(TrackPointer pTrack) {
+    } else {
+        // copy loop in and out points from other deck because any new loops
+        // won't be saved yet
+        m_pLoopInPoint->set(ControlObject::get(
```
For now this is okay, but in the long term I'd like to move away from using ControlObjects to access values from random other places in the code. This makes it difficult to know what code could be affected whenever someone next changes code that impacts the CO.

iamcodemaker replied:
What is the alternative here?

Be-ing commented on src/mixer/basetrackplayer.cpp:
```diff
@@ -279,6 +304,7 @@ void BaseTrackPlayerImpl::slotLoadFailed(TrackPointer pTrack, QString reason) {
     } else {
         qDebug() << "Failed to load track (NULL track object)" << reason;
     }
+    m_cloneFromChannel = NULL;
```
nullptr

Be-ing commented on src/mixer/basetrackplayer.cpp:
```diff
@@ -367,6 +421,8 @@ void BaseTrackPlayerImpl::slotTrackLoaded(TrackPointer pNewTrack,
         qDebug() << "stray BaseTrackPlayerImpl::slotTrackLoaded()";
     }
 
+    m_cloneFromChannel = NULL;
```
nullptr

Be-ing commented on src/mixer/basetrackplayer.cpp:
```diff
@@ -376,6 +432,30 @@ TrackPointer BaseTrackPlayerImpl::getLoadedTrack() const {
+        return;
+    }
+
+    TrackPointer pTrack = m_cloneFromChannel->getEngineBuffer()->getLoadedTrack();
```
The `m_cloneFromChannel` name reads more like a function name than a variable name to me. How about renaming this variable to `m_channelToCloneFrom`?

Be-ing commented on src/mixer/basetrackplayer.cpp:
```diff
@@ -343,22 +369,50 @@ void BaseTrackPlayerImpl::slotTrackLoaded(TrackPointer pNewTrack,
+            // perform a clone of the given channel
+
+            // copy rate
+            if (m_pRateSlider != NULL) {
```
nullptr

Be-ing commented on src/mixer/basetrackplayer.cpp:
```diff
@@ -376,6 +432,30 @@ TrackPointer BaseTrackPlayerImpl::getLoadedTrack() const {
+
+    TrackPointer pTrack = m_cloneFromChannel->getEngineBuffer()->getLoadedTrack();
+    if (!pTrack) {
+        m_cloneFromChannel = NULL;
```
nullptr

Be-ing commented on src/mixer/playermanager.cpp:
```diff
@@ -559,6 +559,17 @@ Auxiliary* PlayerManager::getAuxiliary(unsigned int auxiliary) const {
+void PlayerManager::slotCloneDeck(QString source_group, QString target_group) {
+    BaseTrackPlayer* pPlayer = getPlayer(target_group);
+
+    if (pPlayer == NULL) {
```
nullptr

Be-ing commented on src/util/dnd.h:
```diff
@@ -96,6 +97,29 @@ class DragAndDropHelper {
+    static bool allowDeckCloneAttempt(const QDropEvent& event,
+                                   const QString& group) {
+        // only allow clones to decks
+        if (!PlayerManager::isDeckGroup(group, NULL)) {
```
nullptr

Be-ing commented on src/widget/wtracktext.cpp:
```diff
@@ -56,12 +56,18 @@ void WTrackText::dragEnterEvent(QDragEnterEvent *event) {
 
 void WTrackText::dropEvent(QDropEvent *event) {
     if (DragAndDropHelper::allowLoadToPlayer(m_pGroup, m_pConfig)) {
```
Correct me if I'm wrong, but it looks like the exact same code is duplicated in several widgets. Somehow this should be centralized to one location. How about making a simple TrackDropTarget class that only implements a dropEvent method and making all these widgets inherit from it?

iamcodemaker replied:
I made a `WTrackDropTarget` class and tried to make stuff inherit from it, but it causes there to be multiple `QWidget` classes in the inheritance tree. I haven't tried it with virtual inheritance yet which might fix that, but it might also be easier just to implement a `DragAndDropHelper::trackDropEvent()` function and just call that from  `dropEvent()` (and a similar helper for `dragEnterEvent()`).

Be-ing replied:
Okay, as long as the code isn't duplicated. :+1:

Be-ing reviewed (changes requested):
The code looks pretty good. Mostly small nitpicks in the comments below.

iamcodemaker committed:
rename requestSyncPosition to requestClonePosition

iamcodemaker committed:
rename EngineBuffer::m_pSyncFromChannel to m_pChannelToSyncFrom

iamcodemaker committed:
rename BaseTrackPlayerImpl::m_syncFromChannel to m_pChannelToCloneFrom

iamcodemaker committed:
use nullptr instead of NULL

Be-ing commented on src/engine/enginebuffer.cpp:
```diff
@@ -1121,6 +1129,10 @@ void EngineBuffer::processSlip(int iBufferSize) {
 void EngineBuffer::processSyncRequests() {
+    EngineChannel* pChannel = m_pSyncFromChannel.fetchAndStoreRelaxed(NULL);
+    if (pChannel) {
+        seekCloneBuffer(pChannel->getEngineBuffer());
```
@daschuer what do you think is the right place for this code?

iamcodemaker replied:
I moved this to `EngineBuffer::processSeek()`, it needs to come before the seek operation, and the clone operation is more seek related than sync related so this makes sense to me.

iamcodemaker committed:
use qt5 function pointer syntax for signal/slot connections

iamcodemaker committed:
don't allow deck clone on drag and drop if shift is pressed

iamcodemaker committed:
use nullptr instead of NULL

iamcodemaker committed:
use nullptr instead of NULL

iamcodemaker committed:
use nullptr instead of NULL

iamcodemaker commented:
- [x] Resolve merge conflicts with master
- [x] Make cloning the default drag-and-drop action. Shift + drop clones everything except the play button state.
- [ ] Add [ChannelX], CloneToDeck or [ChannelX], CloneFromDeck ControlObject -- we still need to decide which to implement.

ronso0 commented:
`[ChannelX], CloneToDeckY` :+1:

ronso0 commented:
Great! We need a Clone icon for d'n'd actions so users see the difference when holding Shift.

iamcodemaker committed:
mock clone slots in autodjprocessor_test.cpp

iamcodemaker commented:
After thinking about this more, I think CloneFromDeckX where you specify the deck and a general CloneFromDeck where Mixxx selects the deck would work best.

Be-ing commented:
Sorry if I wasn't clear before. CloneFromDeck wouldn't be a normal binary button. Setting [ChannelX], CloneFromDeck to a value of Y would clone from deck Y to deck X.

iamcodemaker commented:
> Sorry if I wasn't clear before. CloneFromDeck wouldn't be a normal binary button. Setting [ChannelX], CloneFromDeck to a value of Y would clone from deck Y to deck X. I think dedicated CloneFromDeckY buttons would be superfluous.

We are on the same page. I wasn't sure if you could have a control object where you set the channel value like that.

Be-ing commented:
Sure, it would just be a plain ControlObject rather than a ControlPushButton.

daschuer commented:
Ok, go ahead. We just need to decide how and if we integrate Samplers.
Cloning to a sampler is dangerous because no access to certain controls.

iamcodemaker commented:
Is there a use case for cloning to a sampler? I can't think of one. We could just forbid it.

iamcodemaker commented:
Is there a way to write a string to a ControlObject? I only see double values. Right now I'm implementing this with the double meaning the deck number to clone from and 0 meaning automatic selection.

iamcodemaker committed:
add CloneFromDeck control object

Accepts the deck number to clone. If 0 or a negative value is passed
pickNonSyncSyncTarget() is used to choose the clone source.

iamcodemaker commented:
I would use a string to pass `[Channel1]` or whatever as the clone source. Obviously I implemented it without that though. Still planning on addressing the duplicate code for the drop target, there are like 5 duplicated cases and one additional case that can probably be factored out.

Be-ing commented on src/mixer/basetrackplayer.cpp:
```diff
@@ -440,7 +448,20 @@ void BaseTrackPlayerImpl::slotCloneDeck(const QString& group) {
+    if (deck < 1) {
+        slotCloneChannel(m_pEngineMaster->getEngineSync()->pickNonSyncSyncTarget(m_pChannel));
+    } else {
+        slotCloneDeck(PlayerManager::groupForDeck(deck-1));
```
I think rounding to an int is needed here.

iamcodemaker committed:
round value passed to CloneFromDeck

iamcodemaker committed:
factor out track drag and drop code

This code was duplicated in a number of classes, now they all call a
helper funciton that implements the logic. I attempted doing this with
inhertiance, but I couldn't work around QWidget being inherited multiple
times. This solution works just as well for eliminating code
duplication.

iamcodemaker committed:
move clone code to EngineBuffer::processSeek()

At this point of the clone process we need to seek our buffer to the
position of the target channel's buffer so this code is more seek
related than sync related.

Be-ing commented:
Great work, thanks for implementing this long requested feature! I am curious what creative uses people will make of it.

uklotzde commented:
Why did this get merged? Now the tests are broken, i.e. _EngineBufferE2ETest.SoundTouchCrashTest_ sleeps forever in BaseSignalPathTest::loadTrack():

```
0x00007ffff55104c0 in nanosleep () from /usr/lib64/libpthread.so.0
(gdb) bt
#0  0x00007ffff55104c0 in nanosleep () at /usr/lib64/libpthread.so.0
#1  0x00007ffff6e2aa4a in QTest::qSleep(int) () at /usr/lib64/libQt5Test.so.5
#2  0x00000000005fab1a in BaseSignalPathTest::loadTrack(Deck*, std::shared_ptr<Track>) (pTrack=..., pDeck=<optimized out>, this=0x4b16540)
    at src/test/signalpathtest.h:131
#3  0x00000000005fab1a in SignalPathTest::SignalPathTest()
    (this=this@entry=0x4b16540) at src/test/signalpathtest.h:233
#4  0x00000000005faedb in EngineBufferE2ETest::EngineBufferE2ETest()
    (this=0x4b16540) at src/test/enginebuffertest.cpp:204
#5  0x00000000005faedb in EngineBufferE2ETest_SoundTouchCrashTest_Test::EngineBufferE2ETest_SoundTouchCrashTest_Test() (this=0x4b16540)
    at src/test/enginebuffertest.cpp:204
#6  0x00000000005faedb in testing::internal::TestFactoryImpl<EngineBufferE2ETest_SoundTouchCrashTest_Test>::CreateTest() (this=<optimized out>)
    at lib/gtest-1.7.0/include/gtest/internal/gtest-internal.h:454
#7  0x0000000000ba6d1a in testing::internal::HandleSehExceptionsInMethodIfSupported<testing::internal::TestFactoryBase, testing::Test*>(testing::internal::TestFactoryBase*, testing::Test* (testing::internal::TestFactoryBase::*)(), char const*)
    (location=0xeffa68 "the test fixture's constructor", method=<optimized out>, object=0x10ea560) at lib/gtest-1.7.0/src/gtest.cc:2059
#8  0x0000000000ba6d1a in testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::TestFactoryBase, testing::Test*>(testing::internal::TestFact
oryBase*, testing::Test* (testing::internal::TestFactoryBase::*)(), char const*) (object=0x10ea560, method=<optimized out>, location=location@entry=0xeffa68 "the test fixture's constructor") at lib/gtest-1.7.0/src/gtest.cc:2114
#9  0x0000000000b9f433 in testing::TestInfo::Run() (this=0x10ea7e0)
    at lib/gtest-1.7.0/src/gtest.cc:2317
#10 0x0000000000b9f433 in testing::TestInfo::Run() (this=0x10ea7e0)
    at lib/gtest-1.7.0/src/gtest.cc:2300
#11 0x0000000000b9f592 in testing::TestCase::Run() (this=0x10ea930)
    at lib/gtest-1.7.0/src/gtest.cc:2444
#12 0x0000000000b9f592 in testing::TestCase::Run() (this=0x10ea930)
    at lib/gtest-1.7.0/src/gtest.cc:2429
#13 0x0000000000b9febc in testing::internal::UnitTestImpl::RunAllTests()
    (this=0x10da0f0) at /usr/include/c++/8/bits/stl_vector.h:930
#14 0x0000000000b9febc in testing::internal::UnitTestImpl::RunAllTests()
    (this=0x10da0f0) at lib/gtest-1.7.0/src/gtest.cc:4222
#15 0x0000000000ba025b in testing::internal::HandleSehExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*)
    (location=0xeffb30 "auxiliary test code (environments or event listeners)", method=<optimized out>, object=<optimized out>)
    at lib/gtest-1.7.0/src/gtest.cc:2059
#16 0x0000000000ba025b in testing::internal::HandleExceptionsInMethodIfSupported<testing::internal::UnitTestImpl, bool>(testing::internal::UnitTestImpl*, bool (testing::internal::UnitTestImpl::*)(), char const*)
    (location=0xeffb30 "auxiliary test code (environments or event listeners)", method=(bool (testing::internal::UnitTestImpl::*)(testing::internal::UnitTestImpl * const)) 0xb9f640 <testing::internal::UnitTestImpl::RunAllTests()>, object=0x10da0f0) at lib/gtest-1.7.0/src/gtest.cc:2114
#17 0x0000000000ba025b in testing::UnitTest::Run()
    (this=0x107e180 <testing::UnitTest::GetInstance()::instance>)
    at lib/gtest-1.7.0/src/gtest.cc:3926
#18 0x00000000004f055a in RUN_ALL_TESTS() ()
    at lib/gtest-1.7.0/include/gtest/gtest.h:2288
#19 0x00000000004f055a in main(int, char**)
    (argc=<optimized out>, argv=0x7fffffffd448) at src/test/main.cpp:31
```

All platforms seem to be affected.

iamcodemaker commented:
This is a shot in the dark, but this may be happening due to the "clone on double load" logic. Maybe that should be moved to LoadSelectedTrack control object. I'll take a look.

Be-ing commented:
Sorry I did not look much into the tests timing out considering that is a regular occurrence with our CI systems anyway. Thanks for looking into this @iamcodemaker.

iamcodemaker commented:
My hunch was right. Commenting the time based clone trigger in `BaseTrackPlayerImpl::slotLoadTrack()` stops the timeouts. So this should only affect the tests. I'll work on a fix or work around.

iamcodemaker commented:
Fixed in  #2024.

esbrandt commented:
@iamcodemaker 
Thanks for adding that feature.

Trying to incorporate this into my setup, and some questions arise about how this feature is supposed to work.
* Cloning works by dragging a loaded track from one deck to another, while dragging a track from the library to the deck does not invoke cloning, correct?
* Using the ``CloneFromDeck`` control, should it be possible to clone a deck by midi/keyboard shortcut? I added a shortcut for cloning Deck 1, and it does not work as intended.

```bash
Debug [Main]: keyboard press:  "Ctrl+Shift+Left"
Critical [Main]: DEBUG ASSERT: "pBehavior" in function void ControlDoublePrivate::setValueFromMidi(MidiOpCode, double) at src/control/control.cpp:244
Warning [Main]: Cannot set "[Channel1]" , "CloneFromDeck" by Midi
```

Any hints? Thanks

iamcodemaker commented:
Dragging from the library should not trigger a clone. I didn't test this with a keyboard shortcut, it worked from my midi mapping. I'll test with a keyboard shortcut. The other way to clone is to press load twice quickly (`shift+left` or `shift+right`)

daschuer commented:
"[Channel1]" , "CloneFromDeck" is just an ControlObject and expects the 1 based source deck number as value. I think you cannot map it via keyboard. 
For this we need a new ControlPushButton. 
What should happen if a user "Ctrl+Shift+Left"? 
I think we need "[Channel1]" , "CloneFromDeck1" for this. Since we will have  "Ctrl+Shift+Left" and  "Ctrl+Shift+Right", how can we make it more flexible and predicting what the user wants in a four deck layout? 
Ideas?

iamcodemaker commented:
So right now that's the idea behind the double load functionality. It clones the first playing deck. That doesn't work if you want to have two separate groups of doubling going on though or if you have three or 4 decks playing, but for most uses, just cloning the other playing deck will be sufficient.

We could also just hard code cloning between decks 1+2 and 3+4, such that if the user activates clone on one of those decks, the target is always the other deck in the group. Not sure if that's better or not. Personally I like it the way it is now, I think that's more flexible.

ronso0 commented:
One of my Load buttons is faulty (sends multiple 'pressed' signals sometimes). The connected long-press function (set star rating) fails and I'd try to press again, but currently the hard-coded clone functionality gives me a deck clone and I'd have to pick and load the previous track again manually.

Can I somehow opt out of that 'double-press to Load' behaviour?

ronso0 commented:
Maybe it doesn't need an option but just an improvement of the double-press detection.

as described, by accident this press pattern happens with my broken button:
`[looong press] [short release] [short press]`
This unsual click gesture that's not used anywhere else AFAICT and is kinda hard to execute on purpose so I think it only occurs with faulty hardware and can safely be excluded to trigger cloning.


Only trigger cloning when a simple double-tap is detected. This would be an improvement for some, while noone would suffer.
`[short press] [short release] [short press]`

iamcodemaker commented:
I don't think it makes sense to compensate for broken hardware in the code like that. this could have unforseen consequences on non-broken haedware. IMO the best solution would be to detect this in your mapping script and supress the second press. That or patch your local copy.

ronso0 commented:
Okay, I'll find a way around, probably with complex timer filters in my script (as I have a long-press function already).
Though I really think other affected users won't find a way. A (opt-out) Preferences option would still be my preferred choice.

After some quick investigation I think the issue is that the time between button **releases** is measured, rather in between button **presses**, or better: duration of both presses.

Out of curiosity (and to not just cut off the disccussion):
> This could have unforseen consequences on non-broken hardware

I suggest to filter out [long press] [short press] which is a gesture not supposed to do anything, [short press] [short press] would work as before.
What could be potential impairments in your opinion?

iamcodemaker commented:
The control object doesn't know anything about the duration of button presses or the difference between press and release. It just knows a load operation was requested. It's checking the timing between load operations. Thinking about this further, the mapping script is the only place you could implement your solution.

Regarding potential unforseen issues. I don't know, but that's the point. If we are making assumptions about what the user or hardware is telling us (beyond the known expected behavior), there is the potential for our assumptions to be wrong and lead to unexpected behavior. Think of when auto correct fails on your smartphone keyboard.

I think other users with broken hardware will just replace or repair their broken hardware. If every one of these devices was faulty in the same exact way, then it might make sense to distribute this work around in the mapping script.

daschuer commented:
In general, I think the mapping is a actually the right place to fix it. I consider a missing denouncing in the controller firmware more likely than a hardware issue.   

However, something else might be wrong. 
Do you use the "LoadSelectedTrack" control? It is sensitive only for rising slopes. 
[looong press] [short release] [short press]
has two rising slopes. For the clone feature they have to be withing 500 ms. If we assume [looong press] is longer than 500 ms the issue must not happen and we have a bug somewhere else. 

Or do we have actually [looong press] [short release] [short press] [short release] [short press]?

ronso0 commented:
> Or do we have actually [looong press] [short release] [short press] [short release] [short press]?

After further testing I think this is the case unfortuantely. It happened quite often earlier when I tried [looong press] [short press], now I can't reproduce anymore. Maybe because of this extensive button  massage :)
sorry for the noise.

Anyway, the double-tap happens accidentially too often, and filtering this in the script is a lot of work, so I'd prefer this to be opt-out.

daschuer commented:
Ok, this makes sense in this case.
Would you mind to file a bug for reference?

ronso0 commented:
I won't forget as I already work on the peferences option in https://github.com/ronso0/mixxx/tree/clone-deck-opt-out but it'not working yet.

ronso0 commented:
anyone's currently working on the direct-clone COs like `[Channel2],cloneToChannel3`?

iamcodemaker commented:
No, we decided to go the CloneFromDeck route.

ronso0 commented:
I don't mind `to` or `from` if in the end we have 12 different COs to specify the clone route deckX->deckY.
so it's WIP?

iamcodemaker commented:
The CloneFromDeck CO was implemented as part of this PR. You specify the deck number you want to clone from as the parameter.

ronso0 commented:
could you give me a working example? I have a hard time scrolling through the code and expected to find something in controllers/controllerpickermenu.cpp

iamcodemaker commented:
The CloneFromDeck CO is documented on the wiki. I don't have a working example to share. Basically in a controller mapping, it would work like this. The function names and argument order may be wrong, but this is the idea.

```
engine.set("[Channel1]", "CloneFromDeck", 2)
```

That would clone from deck 2 to deck 1.

ronso0 commented:
> `engine.set("[Channel1]", "CloneFromDeck", 2)`

Thanks, I'll try it!

ronso0 commented:
this works:
engine.**setValue**("[Channel1]", "CloneFromDeck", 2)

used in a script:
`engine.setValue( group, "CloneFromDeck", targetDeck);`

esbrandt commented:
Double pressing ``Shift+Left`` and ``Shift+Right`` invoke 2 operations: loading a track to the respective deck, then replacing the just loaded track with the clone. 
  **Issues**
   - Unlike other standard keyboard shortcuts like ``D``or ``L`` , the  hardcoded cloning shortcuts do only work, if the library table has focus.
   -  If the library table has focus, and the highlighted track in a library table is inaccessible, a modal window appears (error message), and cloning stops. 

If the shortcuts for loading and cloning could be decoupled, it would make this feature even more useful. And it makes #2042 obsolete, introducing another preference option just to disable the this feature not necessary.

iamcodemaker commented:
The problem with only having a separate way to trigger this is that you then have to duplicate that everywhere you might trigger a load or clone.

We need a way to say "load button was pressed, then pressed again, cancel the first load". That would solve your second issue and result in the dialog briefly appearing then disappearing.

For first issue, I don't know Mixxx or Qt well enough to really address that. In theory it should be possible to capture any shift+left/right for the whole window and then determine the correct action.

ronso0 commented:
some thoughts I had while working on #2042 : 
can we start a timer on first Load press that waits 500ms?
- if a 2nd Load press happens within that wait period we have a double-tap
- no 2nd Load press: regular load action (with error message if track can't be loaded)

Wouldn't that solve both issues?

ronso0 commented:
#2042 would still be very helpful for controller use case IMO

ronso0 commented:
cloning comes in very handy when I play with 3 decks and want to get the running tracks back to decks 1 & 2.
:+1:

iamcodemaker commented:
> can we start a timer on first Load press that waits for 500ms (current double-tap period)?

I considered this approach. This would work, but we would need to take special care to skip the delay in the "LoadAndPlay" case. Also, a delay would make loading tracks feel laggy.

ronso0 commented:
what about this:
clone a deck by default if it is already **loaded & playing** in another deck?
